In [832]:
import pandas as pd
import numpy as np
import datetime as dt

# read in all our data
df=pd.read_csv("ABDATA.csv")

In [833]:
df.head()

candidateID  citycode  cdi(city development index) gender  \
0         8949  city_103                        0.920   Male   
1        29725   city_40                        0.776   Male   
2        11561   city_21                        0.624    NaN   
3        33241  city_115                        0.789    NaN   
4          666  city_162                        0.767   Male   

       relevent_experience enrollment status (higher education)  \
0  Has relevent experience                        no_enrollment   
1   No relevent experience                        no_enrollment   
2   No relevent experience                     Full time course   
3   No relevent experience                                  NaN   
4  Has relevent experience                        no_enrollment   

  highest qualification   specialization experience (years)  \
0              Graduate             STEM                >20   
1              Graduate             STEM                 15   
2              Graduate             STEM                  5   
3              Graduate  Business Degree                 <1   
4               Masters             STEM                >20   

  companysize (people)    company_type last_new_job (years since)  \
0                  NaN             NaN                          1   
1                50-99         Pvt Ltd                         >4   
2                  NaN             NaN                      never   
3                  NaN         Pvt Ltd                      never   
4                50-99  Funded Startup                          4   

   training_hours(total in current job)  looking to change jobs  
0                                    36                       1  
1                                    47                       0  
2                                    83                       0  
3                                    52                       1  
4                                     8                       0

Data Types of All the Columns

In [834]:
df.dtypes

candidateID                               int64
citycode                                 object
cdi(city development index)             float64
gender                                   object
relevent_experience                      object
enrollment status (higher education)     object
highest qualification                    object
specialization                           object
experience (years)                       object
companysize (people)                     object
company_type                             object
last_new_job (years since)               object
training_hours(total in current job)      int64
looking to change jobs                    int64
dtype: object

Dropping Duplicates if any

In [835]:
df = df.drop_duplicates()
#There was no effect since there was no duplication in the data though that is rarely the case

In [836]:
df

candidateID  citycode  cdi(city development index) gender  \
0             8949  city_103                        0.920   Male   
1            29725   city_40                        0.776   Male   
2            11561   city_21                        0.624    NaN   
3            33241  city_115                        0.789    NaN   
4              666  city_162                        0.767   Male   
...            ...       ...                          ...    ...   
19153         7386  city_173                        0.878   Male   
19154        31398  city_103                        0.920   Male   
19155        24576  city_103                        0.920   Male   
19156         5756   city_65                        0.802   Male   
19157        23834   city_67                        0.855    NaN   

           relevent_experience enrollment status (higher education)  \
0      Has relevent experience                        no_enrollment   
1       No relevent experience                        no_enrollment   
2       No relevent experience                     Full time course   
3       No relevent experience                                  NaN   
4      Has relevent experience                        no_enrollment   
...                        ...                                  ...   
19153   No relevent experience                        no_enrollment   
19154  Has relevent experience                        no_enrollment   
19155  Has relevent experience                        no_enrollment   
19156  Has relevent experience                        no_enrollment   
19157   No relevent experience                        no_enrollment   

      highest qualification   specialization experience (years)  \
0                  Graduate             STEM                >20   
1                  Graduate             STEM                 15   
2                  Graduate             STEM                  5   
3                  Graduate  Business Degree                 <1   
4                   Masters             STEM                >20   
...                     ...              ...                ...   
19153              Graduate       Humanities                 14   
19154              Graduate             STEM                 14   
19155              Graduate             STEM                >20   
19156           High School              NaN                 <1   
19157        Primary School              NaN                  2   

      companysize (people)    company_type last_new_job (years since)  \
0                      NaN             NaN                          1   
1                    50-99         Pvt Ltd                         >4   
2                      NaN             NaN                      never   
3                      NaN         Pvt Ltd                      never   
4                    50-99  Funded Startup                          4   
...                    ...             ...                        ...   
19153                  NaN             NaN                          1   
19154                  NaN             NaN                          4   
19155                50-99         Pvt Ltd                          4   
19156              500-999         Pvt Ltd                          2   
19157                  NaN             NaN                          1   

       training_hours(total in current job)  looking to change jobs  
0                                        36                       1  
1                                        47                       0  
2                                        83                       0  
3                                        52                       1  
4                                         8                       0  
...                                     ...                     ...  
19153                                    42                       1  
19154                                    52                       1  
19155                    

Checking for Missing Values in each of the column

In [837]:
df=df.replace(r'^\s*$', np.nan, regex=True)
missing_values_count = df.isnull().sum()
missing_values_count

candidateID                                0
citycode                                   0
cdi(city development index)                0
gender                                  4508
relevent_experience                        0
enrollment status (higher education)     386
highest qualification                    460
specialization                          2813
experience (years)                        65
companysize (people)                    5938
company_type                            6140
last_new_job (years since)               423
training_hours(total in current job)       0
looking to change jobs                     0
dtype: int64

Checking for Unique Values in Each Column

In [838]:
df.nunique()

candidateID                             19158
citycode                                  123
cdi(city development index)                93
gender                                      3
relevent_experience                         2
enrollment status (higher education)        3
highest qualification                       5
specialization                              6
experience (years)                         22
companysize (people)                        8
company_type                                6
last_new_job (years since)                  6
training_hours(total in current job)      241
looking to change jobs                      2
dtype: int64

# Gender Column

Exploring the unique values in Gender Column

In [839]:
#Fixing hidden repetition of data
explore_values = df[['gender']]
#Step 1 -Recall the replacements needed to be made
for col in explore_values:
    print(explore_values[col].unique())

['Male' nan 'Female' 'Other']


Replaced missing values in gender with 'prefer not to say'

In [840]:
df['gender']=df['gender'].replace(np.nan, 'Prefer not to Say')

Values count of each variables in gender column

In [841]:
df['gender'].value_counts()

Male                 13221
Prefer not to Say     4508
Female                1238
Other                  191
Name: gender, dtype: int64

In [842]:
#Fixing hidden repetition of data
explore_values = df[['gender']]
#Step 1 -Recall the replacements needed to be made
for col in explore_values:
    print(explore_values[col].unique())

['Male' 'Prefer not to Say' 'Female' 'Other']


In [843]:
df.head()

candidateID  citycode  cdi(city development index)             gender  \
0         8949  city_103                        0.920               Male   
1        29725   city_40                        0.776               Male   
2        11561   city_21                        0.624  Prefer not to Say   
3        33241  city_115                        0.789  Prefer not to Say   
4          666  city_162                        0.767               Male   

       relevent_experience enrollment status (higher education)  \
0  Has relevent experience                        no_enrollment   
1   No relevent experience                        no_enrollment   
2   No relevent experience                     Full time course   
3   No relevent experience                                  NaN   
4  Has relevent experience                        no_enrollment   

  highest qualification   specialization experience (years)  \
0              Graduate             STEM                >20   
1              Graduate             STEM                 15   
2              Graduate             STEM                  5   
3              Graduate  Business Degree                 <1   
4               Masters             STEM                >20   

  companysize (people)    company_type last_new_job (years since)  \
0                  NaN             NaN                          1   
1                50-99         Pvt Ltd                         >4   
2                  NaN             NaN                      never   
3                  NaN         Pvt Ltd                      never   
4                50-99  Funded Startup                          4   

   training_hours(total in current job)  looking to change jobs  
0                                    36                       1  
1                                    47                       0  
2                                    83                       0  
3                                    52                       1  
4                                     8                       0

Replacing nan values with mv to deal with missing values easily 

In [844]:
df= df.replace(np.nan, 'mv')
df

candidateID  citycode  cdi(city development index)             gender  \
0             8949  city_103                        0.920               Male   
1            29725   city_40                        0.776               Male   
2            11561   city_21                        0.624  Prefer not to Say   
3            33241  city_115                        0.789  Prefer not to Say   
4              666  city_162                        0.767               Male   
...            ...       ...                          ...                ...   
19153         7386  city_173                        0.878               Male   
19154        31398  city_103                        0.920               Male   
19155        24576  city_103                        0.920               Male   
19156         5756   city_65                        0.802               Male   
19157        23834   city_67                        0.855  Prefer not to Say   

           relevent_experience enrollment status (higher education)  \
0      Has relevent experience                        no_enrollment   
1       No relevent experience                        no_enrollment   
2       No relevent experience                     Full time course   
3       No relevent experience                                   mv   
4      Has relevent experience                        no_enrollment   
...                        ...                                  ...   
19153   No relevent experience                        no_enrollment   
19154  Has relevent experience                        no_enrollment   
19155  Has relevent experience                        no_enrollment   
19156  Has relevent experience                        no_enrollment   
19157   No relevent experience                        no_enrollment   

      highest qualification   specialization experience (years)  \
0                  Graduate             STEM                >20   
1                  Graduate             STEM                 15   
2                  Graduate             STEM                  5   
3                  Graduate  Business Degree                 <1   
4                   Masters             STEM                >20   
...                     ...              ...                ...   
19153              Graduate       Humanities                 14   
19154              Graduate             STEM                 14   
19155              Graduate             STEM                >20   
19156           High School               mv                 <1   
19157        Primary School               mv                  2   

      companysize (people)    company_type last_new_job (years since)  \
0                       mv              mv                          1   
1                    50-99         Pvt Ltd                         >4   
2                       mv              mv                      never   
3                       mv         Pvt Ltd                      never   
4                    50-99  Funded Startup                          4   
...                    ...             ...                        ...   
19153                   mv              mv                          1   
19154                   mv              mv                          4   
19155                50-99         Pvt Ltd                          4   
19156              500-999         Pvt Ltd                          2   
19157                   mv              mv                          1   

       training_hours(total in current job)  looking to change jobs  
0                                        36                       1  
1                                        47                       0  
2                                        83                       0  
3                                        52                       1  
4                                         8                       0  
...                                     ...                     ...  
19153                

Checking for null values in specialization, highest qualification, comapany size and company type 

In [845]:
df[(df['specialization'] ==  'mv') & (df['highest qualification'] ==  'mv') & (df['companysize (people)'] ==  'mv') & (df['company_type'] ==  'mv')].count()

candidateID                             278
citycode                                278
cdi(city development index)             278
gender                                  278
relevent_experience                     278
enrollment status (higher education)    278
highest qualification                   278
specialization                          278
experience (years)                      278
companysize (people)                    278
company_type                            278
last_new_job (years since)              278
training_hours(total in current job)    278
looking to change jobs                  278
dtype: int64

Checking for null values in  last new job,comapany size and company type

In [846]:
df[ (df['companysize (people)'] == 'mv') & (df['company_type'] =='mv') & (df['last_new_job (years since)'] == 'mv')].count()

candidateID                             193
citycode                                193
cdi(city development index)             193
gender                                  193
relevent_experience                     193
enrollment status (higher education)    193
highest qualification                   193
specialization                          193
experience (years)                      193
companysize (people)                    193
company_type                            193
last_new_job (years since)              193
training_hours(total in current job)    193
looking to change jobs                  193
dtype: int64

Dropping the null values in specialization, highest qualification, comapany size and company type, as no pattern exist to impute the values

In [847]:
df.drop(df[(df['specialization'] == 'mv') & (df['highest qualification'] == 'mv') & (df['companysize (people)'] == 'mv') & (df['company_type'] == 'mv')].index, inplace = True)

Dropping the null values in last new job, highest qualification, comapany size and company type, as no pattern exist to impute the values

In [848]:
df.drop(df[(df['last_new_job (years since)'] == 'mv') & (df['companysize (people)'] == 'mv') & (df['company_type'] == 'mv')].index, inplace = True)

In [849]:
df

candidateID  citycode  cdi(city development index)             gender  \
0             8949  city_103                        0.920               Male   
1            29725   city_40                        0.776               Male   
2            11561   city_21                        0.624  Prefer not to Say   
3            33241  city_115                        0.789  Prefer not to Say   
4              666  city_162                        0.767               Male   
...            ...       ...                          ...                ...   
19153         7386  city_173                        0.878               Male   
19154        31398  city_103                        0.920               Male   
19155        24576  city_103                        0.920               Male   
19156         5756   city_65                        0.802               Male   
19157        23834   city_67                        0.855  Prefer not to Say   

           relevent_experience enrollment status (higher education)  \
0      Has relevent experience                        no_enrollment   
1       No relevent experience                        no_enrollment   
2       No relevent experience                     Full time course   
3       No relevent experience                                   mv   
4      Has relevent experience                        no_enrollment   
...                        ...                                  ...   
19153   No relevent experience                        no_enrollment   
19154  Has relevent experience                        no_enrollment   
19155  Has relevent experience                        no_enrollment   
19156  Has relevent experience                        no_enrollment   
19157   No relevent experience                        no_enrollment   

      highest qualification   specialization experience (years)  \
0                  Graduate             STEM                >20   
1                  Graduate             STEM                 15   
2                  Graduate             STEM                  5   
3                  Graduate  Business Degree                 <1   
4                   Masters             STEM                >20   
...                     ...              ...                ...   
19153              Graduate       Humanities                 14   
19154              Graduate             STEM                 14   
19155              Graduate             STEM                >20   
19156           High School               mv                 <1   
19157        Primary School               mv                  2   

      companysize (people)    company_type last_new_job (years since)  \
0                       mv              mv                          1   
1                    50-99         Pvt Ltd                         >4   
2                       mv              mv                      never   
3                       mv         Pvt Ltd                      never   
4                    50-99  Funded Startup                          4   
...                    ...             ...                        ...   
19153                   mv              mv                          1   
19154                   mv              mv                          4   
19155                50-99         Pvt Ltd                          4   
19156              500-999         Pvt Ltd                          2   
19157                   mv              mv                          1   

       training_hours(total in current job)  looking to change jobs  
0                                        36                       1  
1                                        47                       0  
2                                        83                       0  
3                                        52                       1  
4                                         8                       0  
...                                     ...                     ...  
19153                

Total of 443 rows dropped

# Highest Qualification Column

In [850]:

func = lambda x: round(100*x.count()/df.shape[0],1)
pd.pivot_table(df, index='highest qualification' , values='candidateID', aggfunc=func,
                         margins=True,
                         margins_name="Total")

candidateID
highest qualification             
Graduate                      61.5
High School                   10.5
Masters                       23.2
Phd                            2.2
Primary School                 1.6
mv                             1.0
Total                        100.0

Pivot table shows that 1% missing values existin this column and mode is Graduate which is 61.5% 
It is feasible to replace missing values with Graduate

In [851]:
explore_values = df[['highest qualification']]

for col in explore_values:
    print(explore_values[col].unique())
df['highest qualification'].value_counts()

['Graduate' 'Masters' 'High School' 'Phd' 'mv' 'Primary School']


Graduate          11510
Masters            4338
High School        1972
Phd                 412
Primary School      301
mv                  182
Name: highest qualification, dtype: int64

In [852]:
df['highest qualification'].mode()

0    Graduate
dtype: object

In [853]:
for index in df.index:
    if df.loc[index, 'highest qualification']  == 'mv' :
        df.loc[index, 'highest qualification'] = 'Graduate'


Missing values Replaced with Graduate

In [854]:
explore_values = df[['highest qualification']]

for col in explore_values:
    print(explore_values[col].unique())
df['highest qualification'].value_counts()

['Graduate' 'Masters' 'High School' 'Phd' 'Primary School']


Graduate          11692
Masters            4338
High School        1972
Phd                 412
Primary School      301
Name: highest qualification, dtype: int64

# Enrollment Status Column

In [855]:
table2 = pd.pivot_table(df,index='highest qualification', columns='enrollment status (higher education)', values='candidateID',aggfunc='count', fill_value=0,
                         margins=True,
                         margins_name="Total")
table2

enrollment status (higher education)  Full time course  Part time course   mv  \
highest qualification                                                           
Graduate                                          2191               790  183   
High School                                        841               163   44   
Masters                                            550               204   51   
Phd                                                 24                12    6   
Primary School                                      32                10    8   
Total                                             3638              1179  292   

enrollment status (higher education)  no_enrollment  Total  
highest qualification                                       
Graduate                                       8528  11692  
High School                                     924   1972  
Masters                                        3533   4338  
Phd                                             370    412  
Primary School                                  251    301  
Total                                         13606  18715

In [856]:
func = lambda x: round(100*x.count()/df.shape[0],1)
pd.pivot_table(df, index='highest qualification',columns='enrollment status (higher education)' , values='candidateID', aggfunc=func,
                         margins=True,
                         margins_name="Total")

enrollment status (higher education)  Full time course  Part time course   mv  \
highest qualification                                                           
Graduate                                          11.7               4.2  1.0   
High School                                        4.5               0.9  0.2   
Masters                                            2.9               1.1  0.3   
Phd                                                0.1               0.1  0.0   
Primary School                                     0.2               0.1  0.0   
Total                                             19.4               6.3  1.6   

enrollment status (higher education)  no_enrollment  Total  
highest qualification                                       
Graduate                                       45.6   62.5  
High School                                     4.9   10.5  
Masters                                        18.9   23.2  
Phd                                             2.0    2.2  
Primary School                                  1.3    1.6  
Total                                          72.7  100.0

Pivot table shows that 1.6% missing values existin this column and mode is no_enrollment which is 72.7% It is feasible to replace missing values with no_enrollment

In [857]:
df['enrollment status (higher education)'].mode()

0    no_enrollment
dtype: object

Missing values Replaced with No Enrollment

In [858]:
for index in df.index:
    if df.loc[index, 'enrollment status (higher education)'] == 'mv' :
        df.loc[index, 'enrollment status (higher education)'] = 'no_enrollment'

           

In [859]:
#Fixing hidden repetition of data
explore_values = df[['enrollment status (higher education)']]
#Step 1 -Recall the replacements needed to be made
for col in explore_values:
    print(explore_values[col].unique())
df['enrollment status (higher education)'].value_counts()

['no_enrollment' 'Full time course' 'Part time course']


no_enrollment       13898
Full time course     3638
Part time course     1179
Name: enrollment status (higher education), dtype: int64

# Specialization Column

Making relationship of Highest Qualification and Specialization

In [860]:
table3 = pd.pivot_table(df,index='highest qualification', columns='specialization', values='candidateID',aggfunc='count', fill_value=0,
                         margins=True,
                         margins_name="Total")
table3

specialization         Arts  Business Degree  Humanities  No Major  Other  \
highest qualification                                                       
Graduate                207              238         419       190    271   
High School               0                0           0         0      0   
Masters                  42               86         223        29    100   
Phd                       3                3          24         0      9   
Primary School            0                0           0         0      0   
Total                   252              327         666       219    380   

specialization          STEM    mv  Total  
highest qualification                      
Graduate               10166   201  11692  
High School                0  1972   1972  
Masters                 3853     5   4338  
Phd                      373     0    412  
Primary School             0   301    301  
Total                  14392  2479  18715

As pivot table shows we have 2479 missing values in total 

In [861]:
func = lambda x: round(100*x.count()/df.shape[0],1)
pd.pivot_table(df, index='highest qualification',columns='specialization' , values='candidateID', aggfunc=func,
                         margins=True,
                         margins_name="Total")

specialization         Arts  Business Degree  Humanities  No Major  Other  \
highest qualification                                                       
Graduate                1.1              1.3         2.2       1.0    1.4   
High School             NaN              NaN         NaN       NaN    NaN   
Masters                 0.2              0.5         1.2       0.2    0.5   
Phd                     0.0              0.0         0.1       NaN    0.0   
Primary School          NaN              NaN         NaN       NaN    NaN   
Total                   1.3              1.7         3.6       1.2    2.0   

specialization         STEM    mv  Total  
highest qualification                     
Graduate               54.3   1.1   62.5  
High School             NaN  10.5   10.5  
Masters                20.6   0.0   23.2  
Phd                     2.0   NaN    2.2  
Primary School          NaN   1.6    1.6  
Total                  76.9  13.2  100.0

As from the pivot table we can analyze that High School and Primary School has no specialization so we can categorized them in No major 

Which leaves us with only 1.1% missing values which can be replaced with STEM as it dominates the column with 76.9 % including missing values. 

In [862]:
for index in df.index:
 #if graduate,phd or masters then stem  
    if (df.loc[index, 'highest qualification']  == 'Graduate' or df.loc[index, 'highest qualification']  == 'Phd' or df.loc[index, 'highest qualification']  == 'Masters') and  df.loc[index, 'specialization'] == 'mv' :
        df.loc[index, 'specialization'] = 'STEM'
 # if high school or primary then no major       
    elif (df.loc[index, 'highest qualification']  == 'High School' or df.loc[index, 'highest qualification']  == 'Primary School') and  df.loc[index, 'specialization'] == 'mv':
        df.loc[index, 'specialization'] = 'No Major'
         

Missing Values Replaced

In [863]:
#Fixing hidden repetition of data
explore_values = df[['specialization']]
#Step 1 -Recall the replacements needed to be made
for col in explore_values:
    print(explore_values[col].unique())
df['specialization'].value_counts()

['STEM' 'Business Degree' 'No Major' 'Arts' 'Humanities' 'Other']


STEM               14598
No Major            2492
Humanities           666
Other                380
Business Degree      327
Arts                 252
Name: specialization, dtype: int64

# experience (years) and last_new_job (years since) Columns

In [864]:
func = lambda x: round(100*x.count()/df.shape[0],1)
pd.pivot_table(df, index='last_new_job (years since)', values='candidateID', aggfunc=func,
                         margins=True,
                         margins_name="Total")

candidateID
last_new_job (years since)             
1                                  42.6
2                                  15.4
3                                   5.5
4                                   5.5
>4                                 17.5
mv                                  1.2
never                              12.3
Total                             100.0

Since we have small number of missing vlaues in this column which are 1.2%, we assumed these to be 'never' 

In [865]:
for index in df.index:
    if df.loc[index, 'last_new_job (years since)']  == 'mv':
        df.loc[index, 'last_new_job (years since)'] = 'never'

In [866]:
df['experience (years)'].value_counts()

>20    3250
5      1390
4      1358
3      1288
6      1190
2      1062
7      1014
10      983
9       969
8       791
15      682
11      659
14      584
1       510
16      502
12      487
<1      484
13      398
17      339
19      302
18      276
20      146
mv       51
Name: experience (years), dtype: int64

Since we have only 51 missing values out of 19K vlaues in this column, we assume these candidates have 0 years of experience

In [867]:
for index in df.index:
    if df.loc[index, 'experience (years)']  == 'mv':
        df.loc[index, 'experience (years)'] = 0

# companysize (people) Column

In [868]:
df['companysize (people)'].value_counts()

mv           5495
50-99        3083
100-500      2571
10000+       2019
10/49        1471
1000-4999    1328
<10          1308
500-999       877
5000-9999     563
Name: companysize (people), dtype: int64

As we can see we have one column of 10/49 which in excel represented Oct-49 which is was in date format, which was orignally suppose to be 10-49 because inly that range is missing.
So we replaced the value by 10-49

In [869]:
for index in df.index:
    if df.loc[index, 'companysize (people)']  == '10/49':
        df.loc[index, 'companysize (people)'] = '10-49'

Values Replaced

In [870]:
func = lambda x: round(100*x.count()/df.shape[0],1)
pd.pivot_table(df, index='companysize (people)' , values='candidateID', aggfunc=func,
                         margins=True,
                         margins_name="Total")

candidateID
companysize (people)             
10-49                         7.9
100-500                      13.7
1000-4999                     7.1
10000+                       10.8
50-99                        16.5
500-999                       4.7
5000-9999                     3.0
<10                           7.0
mv                           29.4
Total                       100.0

As we can see we have 29.4% of missing values in this column which is 5495, we can't assume any value and impute in this unless the logic is strong, so we made a new variable called unknown

In [871]:
for index in df.index:
    if df.loc[index, 'companysize (people)']  == 'mv':
        df.loc[index, 'companysize (people)'] = 'Unknown'

In [872]:
df['companysize (people)'].value_counts()

Unknown      5495
50-99        3083
100-500      2571
10000+       2019
10-49        1471
1000-4999    1328
<10          1308
500-999       877
5000-9999     563
Name: companysize (people), dtype: int64

# company_type Column

In [873]:
df['company_type'].value_counts()

Pvt Ltd                9817
mv                     5697
Funded Startup         1001
Public Sector           955
Early Stage Startup     603
NGO                     521
Other                   121
Name: company_type, dtype: int64

In [874]:
func = lambda x: round(100*x.count()/df.shape[0],1)
pd.pivot_table(df, index='company_type' , values='candidateID', aggfunc=func,
                         margins=True,
                         margins_name="Total")

candidateID
company_type                    
Early Stage Startup          3.2
Funded Startup               5.3
NGO                          2.8
Other                        0.6
Public Sector                5.1
Pvt Ltd                     52.5
mv                          30.4
Total                      100.0

As we can see we have 30.4% of missing values in this column which is 5697, we can't assume any value and impute in this unless the logic is strong, so we made a new variable called unknown

In [875]:
for index in df.index:
    if df.loc[index, 'company_type']  == 'mv':
        df.loc[index, 'company_type'] = 'Unknown'

In [876]:
df['company_type'].value_counts()

Pvt Ltd                9817
Unknown                5697
Funded Startup         1001
Public Sector           955
Early Stage Startup     603
NGO                     521
Other                   121
Name: company_type, dtype: int64

# Checking of Missing Values

In [877]:
df= df.replace('mv',np.nan)
df

candidateID  citycode  cdi(city development index)             gender  \
0             8949  city_103                        0.920               Male   
1            29725   city_40                        0.776               Male   
2            11561   city_21                        0.624  Prefer not to Say   
3            33241  city_115                        0.789  Prefer not to Say   
4              666  city_162                        0.767               Male   
...            ...       ...                          ...                ...   
19153         7386  city_173                        0.878               Male   
19154        31398  city_103                        0.920               Male   
19155        24576  city_103                        0.920               Male   
19156         5756   city_65                        0.802               Male   
19157        23834   city_67                        0.855  Prefer not to Say   

           relevent_experience enrollment status (higher education)  \
0      Has relevent experience                        no_enrollment   
1       No relevent experience                        no_enrollment   
2       No relevent experience                     Full time course   
3       No relevent experience                        no_enrollment   
4      Has relevent experience                        no_enrollment   
...                        ...                                  ...   
19153   No relevent experience                        no_enrollment   
19154  Has relevent experience                        no_enrollment   
19155  Has relevent experience                        no_enrollment   
19156  Has relevent experience                        no_enrollment   
19157   No relevent experience                        no_enrollment   

      highest qualification   specialization experience (years)  \
0                  Graduate             STEM                >20   
1                  Graduate             STEM                 15   
2                  Graduate             STEM                  5   
3                  Graduate  Business Degree                 <1   
4                   Masters             STEM                >20   
...                     ...              ...                ...   
19153              Graduate       Humanities                 14   
19154              Graduate             STEM                 14   
19155              Graduate             STEM                >20   
19156           High School         No Major                 <1   
19157        Primary School         No Major                  2   

      companysize (people)    company_type last_new_job (years since)  \
0                  Unknown         Unknown                          1   
1                    50-99         Pvt Ltd                         >4   
2                  Unknown         Unknown                      never   
3                  Unknown         Pvt Ltd                      never   
4                    50-99  Funded Startup                          4   
...                    ...             ...                        ...   
19153              Unknown         Unknown                          1   
19154              Unknown         Unknown                          4   
19155                50-99         Pvt Ltd                          4   
19156              500-999         Pvt Ltd                          2   
19157              Unknown         Unknown                          1   

       training_hours(total in current job)  looking to change jobs  
0                                        36                       1  
1                                        47                       0  
2                                        83                       0  
3                                        52                       1  
4                                         8                       0  
...                                     ...                     ...  
19153                

converting mv back to np.nan to check the missing values

In [878]:
df=df.replace(r'^\s*$', np.nan, regex=True)
missing_values_count = df.isnull().sum()
missing_values_count

candidateID                             0
citycode                                0
cdi(city development index)             0
gender                                  0
relevent_experience                     0
enrollment status (higher education)    0
highest qualification                   0
specialization                          0
experience (years)                      0
companysize (people)                    0
company_type                            0
last_new_job (years since)              0
training_hours(total in current job)    0
looking to change jobs                  0
dtype: int64

In [879]:
df.to_excel(r'BI-Project.xlsx', index = False)